Applicability & Alternatives (#265)
ietf-wg-spice/draft-ietf-spice-sd-cwt: draft-ietf-spice-sd-cwt.md

Reviewer: It seems to me like this statement about traceability is the linchpin on what a lot of the new additions turn.  However, it's not directly connected to that discussion, so some of the later statements appear out of context.  What I suggest you do is add a specific section on the claims being traceable, exploring the consequences of that from there.

Either way, you need to make a better connection between the term "traceable", as used here, and "linkability", as used in later discussion.

Author: I think the intro is not the right place to get into this detail, I will try to expand on it in the link-ability section.

  RFC6973:
  I-D.draft-ietf-oauth-selective-disclosure-jwt: SD-JWT
  I-D.draft-ietf-oauth-sd-jwt-vc: SD-JWT-VC

>   I-D.draft-ietf-oauth-status-list: OAUTH-STATUS-LIST

Author: ```suggestion
```

This document defines a generic container format, not a specific credential type.
For example, a license to operate a vehicle and a license to import a product will contain different attributes.

SD-CWT is especially useful in non-human identity use cases, where a device, application, workload, product, or organization needs to share only part of what it is with another system, without exposing every attribute it carries.
Examples include device inspection records, software supply chain statements, product passports, import or export certificates, and other machine-readable credentials used for safety, compliance, asset tracking, or lifecycle management across organizational boundaries.
In these settings, some linkability is often expected because the Verifier needs to know which device, service, component, or legal entity it is dealing with.
SD-CWT helps reduce unnecessary disclosure while still allowing required identifiers and other mandatory claims to be presented.
It is a less natural fit for use cases where avoiding correlation across presentations is the primary privacy goal.

> This includes cases where the subject should be able to make repeated presentations without different Verifiers, or the same Verifier over time, learning that those presentations came from the same entity.

Reviewer: ```suggestion
Cases that have higher requirements for privacy include where the subject should be able to make repeated presentations without different Verifiers -- or the same Verifier over time -- learning that those presentations came from the same entity.
```

Now, I think that this is not your primary concern for SD-CWT.  In these cases specifically, you can simply use a fresh SD-CWT for each presentation.  The case where SD-CWT is completely unusable is where you don't want the Issuer knowing how you use the credential.

Author: Attempted to address the last sentence in https://github.com/ietf-wg-spice/draft-ietf-spice-sd-cwt/pull/265/commits/910784f053686373259bc77ed7cdc0cc73a385ba

> Use cases that depend on strong unlinkability, one-time pseudonyms, or very low observability are generally better served by mechanisms designed to minimize or remove stable identifiers and presentation-level traceability.

Reviewer 2: How is this slight reframe of one-time credentials?

```suggestion
Generating one-time pseudonyms, each using a unique confirmation key, can somewhat mitigate these concerns, but at the expense of managing a proliferation of credentials.
Use cases that depend on strong unlinkability, or very low observability are generally better served by mechanisms designed to minimize or remove stable identifiers and presentation-level traceability.
```

This document defines a generic container format, not a specific credential type.
For example, a license to operate a vehicle and a license to import a product will contain different attributes.

> SD-CWT-based credential types rely on Verifiers trusting the Issuer. While policy or legal frameworks may limit an Issuer's ability to determine how issued credentials are used, SD-CWT provides no cryptographic mechanism to prevent such tracking.

Reviewer: I'm not sure what this is intended to convey.

The statement about verifiers trusting issuers isn't very well qualified.  Generally speaking, if you are going to act on a credential, any action you take is determined by whether you think that the entity making claims (the Issuer) is plausibly able to make those claims.  Which usually means that a Verifier will have a list of Issuers that they consider to be authoritative for certain sets of claims (either as a list of Issuers who can make claims on a certain topic, closer to the Web PKI mode; or, as a list of claims that each Issuer can make claims for).

The policy/legal frameworks thing seems to be addressing a different point.  What point is that?  What does "such tracking" refer to?

Author: I removed this sentence... It all boils down to what is already documented in the unlinkability section of SD-JWT.

> Consequently, SD-CWT is unsuitable for use cases where preventing the Issuer from learning how credentials are used is a requirement.

Reviewer: If the intent of the previous statement was to say that a Verifier can conspire with an Issuer to track the usage of credentials, then this statement would make sense.  Otherwise, I'm not sure what your intent would be.

Author: Yep, hopefully the revised text is clearer.

> SD-CWT is especially useful in non-human identity use cases, where a device, application, workload, product, or organization needs to share only part of what it is with another system, without exposing every attribute it carries.

Reviewer: This still suffers from [WALLOFTEXT](https://en.uncyclopedia.co/wiki/Wall_of_Text).  Can you break it down a little?

Author: Remove use case stuff in https://github.com/ietf-wg-spice/draft-ietf-spice-sd-cwt/pull/265/commits/d60eb53d1c31ac2b236d38bcc93a080e5ba03be1

Examples include device inspection records, software supply chain statements, product passports, import or export certificates, and other machine-readable credentials used for safety, compliance, asset tracking, or lifecycle management across organizational boundaries.
In these settings, some linkability is often expected because the Verifier needs to know which device, service, component, or legal entity it is dealing with.
SD-CWT helps reduce unnecessary disclosure while still allowing required identifiers and other mandatory claims to be presented.

> It is a less natural fit for use cases where avoiding correlation across presentations is the primary privacy goal.

Reviewer: Introducing "correlation" might confuse things without a section heading under which to group the concepts.

> Cases that have higher requirements for privacy include where the subject should be able to make repeated presentations without different Verifiers -- or the same Verifier over time -- learning that those presentations came from the same entity.

Reviewer: Missing word, maybe a tweak?

```suggestion
Cases that have higher requirements for privacy include where the subject should be able to make repeated presentations without different Verifiers -- or the same Verifier over time -- without revealing that those presentations came from the same entity.
```

Generating one-time pseudonyms, each using a unique confirmation key, can somewhat mitigate these concerns, but at the expense of managing a proliferation of credentials.

> Use cases that depend on strong unlinkability, or very low observability are generally better served by mechanisms designed to minimize or remove stable identifiers and presentation-level traceability.

Reviewer: ```suggestion
Use cases that depend on strong unlinkability or very low observability are generally better served by mechanisms designed to minimize or remove stable identifiers and presentation-level traceability.
```

Just fixing a comma there, but I'm not sure that this statement is going to be helpful.  I think that what you want to say is that in cases where the Issuer is part of the privacy threat model will need a different mechanism entirely.

> Examples include credentials used in systems that could reveal a person's travel patterns, current location, or movement over time, and credentials used to access sensitive locations or services where repeated presentation could expose visits to shelters, clinics, places of worship, political events, or counseling centers.

Reviewer: This is very heavily connected to physical location.  You might consider grouping those threats to safety through revealing a person's physical location as distinct from threats to safety through revealing political activity, associations, use of services, etc...  This is a pretty wide space and you might want to make it clear that these are just some of the things that might need to be protected, in two dimensions: the information that is revealed, and what of that information might lead to a safety risk.

Author: Agreed, I removed this.

That is, SD-CWT does not naturally provide Verifier-Verifier unlinkability.
Verifier-Verifier unlinkability can be achieved by requesting a fresh, uncorrelated copy of the credential for each presentation, at a credential management complexity cost.
This assumes that the revealed claims do not contain information that could enable Verifiers to correlate presentations.
If the claims, or the timing or structure of their presentation, leak correlatable information, then fresh copies of credentials may be insufficient.
SD-CWT is designed to mitigate such leakage; however, implementers must ensure that no leakage occurs (for example, by appropriately using decoy digests, see {{decoys}}).

Any Claim Value that pertains to a sufficiently small set of subjects can be used to facilitate tracking the subject.
For example, a high precision issuance time might match the issuance of only a few credentials for a given Issuer, and as such, any presentation of a credential issued at that time can be determined to be associated with the set of credentials issued at that time, for those subjects.

> 

Reviewer 3: ```suggestion
```

This section describes the privacy considerations in accordance with the recommendations from {{RFC6973}}.
Many of the topics discussed in {{RFC6973}} apply to SD-CWT, but are not repeated here.

> ## Correlation & Linkability

Reviewer 2: ```suggestion
## Correlation and Linkability
```

This section describes the privacy considerations in accordance with the recommendations from {{RFC6973}}.
Many of the topics discussed in {{RFC6973}} apply to SD-CWT, but are not repeated here.

## Correlation & Linkability

The term linkability or unlinkability applies to presentations of SD-CWT in the same ways that it applies to SD-JWT as described in {{Section 10.1 of -SD-JWT}}.

> SD-CWT cannot prevent Issuer/Verifier linkability.

Reviewer: Is it / or -?

```suggestion
SD-CWT cannot prevent Issuer-Verifier linkability.
```

In cases where the Issuer is part of the privacy threat model, a different mechanism than SD-CWT is needed.

Presentations of the same SD-CWT to multiple Verifiers can be correlated by matching on the signature component of the COSE_Sign1.
That is, SD-CWT does not naturally provide Verifier-Verifier unlinkability.
Signature based linkability can be mitigated by issuing multiple single-use tokens, at a credential management complexity cost.

> Any presentation of an SD-CWT can be correlated with its issuance if an Issuer and Verifier cooperate.

Reviewer: May move this up with line 1243.

Additionally, compromise of a Holder's device storage presents a more significant risk in pseudonymous systems. If a device containing pseudonymous credentials is stolen and accessed by an attacker, the attacker can present the credentials to Verifiers without the same constraints that identity-bound credentials impose. Because the credentials do not cryptographically bind to the Holder's real identity, Verifiers lack out-of-band verification mechanisms (such as identity verification) to detect that a compromised device is being used by someone other than the legitimate Holder, unless additional controls such as biometric binding are implemented at the device level.

> See {{-CBOR-WEB-PROOF}} for an alternative to SD-CWT which aims to address these concerns.

Reviewer: I had a bunch of suggestions about what this might say about the claims themselves; was that no good?  Because this is all still stuff about other technology, not SD-CWT.

Some of that was already present, which is fine if you think that's adequate.  Though I did suggest that maybe you could reference the stub/dummy claims section as a way to hide information that leaks through the credential structure and might be linkable as a result.

Author: I got halfway through applying updates, before I got pulled into other stuff. I have some refinement on your text locally that I will push when I am able to tomorrow.

Beyond key compromise, attacks targeting the provisioning and binding between issuer names and their cryptographic key material pose significant risks.
An attacker who can manipulate these bindings could substitute their own keys for legitimate issuer keys, enabling credential forgery while appearing to be a trusted issuer.

## Linkability

The term linkability or unlinkability applies to presentations of SD-CWT in the same ways that it applies to SD-JWT as described in {{Section 10.1 of -SD-JWT}}.

> As noted in the introduction, SD-CWT cannot prevent Issuer/Verifier linkability.

Reviewer: This is not particularly clear in the introduction, as it currently stands.  Maybe just restate this directly:

```suggestion
SD-CWT cannot prevent Issuer/Verifier linkability.
```

> Pseudonymous credential schemes, by contrast, employ mechanisms such as fresh pseudonyms or cryptographic unlinking per presentation to enable Holders to prove attributes or properties without revealing stable identifiers, thereby preventing Issuer-Verifier and Verifier-Verifier linkability across presentations.

Reviewer: The remaining discussion is about options that aren't described in this document, so it doesn't really belong here.  I would focus on the limitations of SD-CWT instead:

* No Issuer-Verifier unlinkability.  If you need that (maybe give some examples, non-exhaustively), this isn't the tool for you.
* Verifier-Verifier unlinkability can be achieved only by requesting a fresh, uncorrelated copy of the credential for each presentation.  This assumes that the revealed claims do not contain information that might allow Verifiers to link information.  If the claims - or the timing or structure of their presentation - leak information, then fresh copies of credentials might be insufficient.  SD-CWT is designed to avoid this leakage, but care is needed to ensure that leakage doesn't occur (consider referencing the dummy claims, etc...)

Author: I think this has been addressed.
